x-pack/plugins/enterprise_search/server/utils/elasticsearch_error_handler.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { RequestHandler } from '@kbn/core/server';
+import { Logger } from '@kbn/core/server';
+
+import { i18n } from '@kbn/i18n';
+
+import { ErrorCode } from '../../common/types/error_codes';
+
+import { createError, EnterpriseSearchError } from './create_error';
+import { isUnauthorizedException } from './identify_exceptions';
+
+export function elasticsearchErrorHandler<ContextType, RequestType, ResponseType>(
+  log: Logger,
+  requestHandler: RequestHandler<ContextType, RequestType, ResponseType>
+): RequestHandler<ContextType, RequestType, ResponseType> {
+  return async (context, request, response) => {
+    try {
+      return await requestHandler(context, request, response);
+    } catch (error) {
+      let enterpriseSearchError: EnterpriseSearchError | undefined;
+
+      if (isUnauthorizedException(error)) {
+        enterpriseSearchError = {
+          errorCode: ErrorCode.UNAUTHORIZED,
+          message: i18n.translate('xpack.enterpriseSearch.server.routes.unauthorizedError', {
+            defaultMessage: 'You do not have sufficient permissions.',
+          }),
+          statusCode: 403,
+        };
+      } else {
+        enterpriseSearchError = {
+          errorCode: ErrorCode.UNCAUGHT_EXCEPTION,
+          message: i18n.translate('xpack.enterpriseSearch.server.routes.uncaughtExceptionError', {
+            defaultMessage: 'Enterprise Search encountered an error.',
+          }),
+          statusCode: 502,
+        };
+      }
```

Comment:
@sphilipse @efegurkan I'd like your opinion on a generic uncaught exception handler like this, we had a lot of index-related server routes that raised an error message like `Error fetching data from Enterprise Search`.  But if we keep those, they'll catch any possible authorization errors from being handled here. Otherwise, those routes on non-authorization error would return a generic `An internal server error occurred. Check Kibana server logs for details` message by the Kibana middlewar (which was the original confusing behavior that spawned this PR in the first place 🤪)

Reply:
Eh....I like the second one ('Check Kibana server logs for details') more because that's more applicable in all of our new routes. Maybe we want to turn that into 'An internal server error occurred. Check the Kibana server logs and the Enterprise Search logs for more details'"?

Reply:
I actually append that over here before the error message is bubbled up https://github.com/elastic/kibana/pull/138341/files#diff-d81b2a9013bbd00a83612420e2b37ae5a6cd4e506a3ad4aa691a30943b368aacR60-R61. I do it here because IMO there's no point in logging "Check your kibana server logs" in the kibana server logs.